### ISSUE TITLE
AttributeError: close

### ISSUE THREAD

**Issue by tk0miya** [blockdiag/blockdiag#69]
Hi,

I have following code:

```
#!python

nwdiag {
  a [shape = "cisco.router"];
  b [shape = "cisco.router_firewall"];
  c [shape = "cisco.pc"];
  a -- b -- c;
}

```

If i run:

```
# nwdiag --debug -T PNG -o neu.png neu
```

... i get:

```
File "/home/victor/work/blog/pelican/env/lib/python2.7/site-packages/blockdiag/drawer.py", line 88, in draw
    self._draw_background()
  File "/home/victor/work/blog/pelican/env/lib/python2.7/site-packages/nwdiag/drawer.py", line 40, in _draw_background
    super(DiagramDraw, self)._draw_background()
  File "/home/victor/work/blog/pelican/env/lib/python2.7/site-packages/blockdiag/drawer.py", line 114, in _draw_background
    shape = r(node, self.metrics)
  File "/home/victor/work/blog/pelican/env/lib/python2.7/site-packages/blockdiagcontrib/cisco.py", line 38, in __init__
    size = images.get_image_size(image_path)
  File "/home/victor/work/blog/pelican/env/lib/python2.7/site-packages/blockdiag/utils/images.py", line 52, in get_image_size
    f.close()
  File "/home/victor/work/blog/pelican/env/lib/python2.7/site-packages/PIL/Image.py", line 537, in __getattr__
    raise AttributeError(name)
AttributeError: close
```

The main problem seems to be in `blockdiag/utils/images.py' in the `get_image_size` function:

```
def get_image_size(filename):
    image = None
    try:
        image = open(filename)
        return image.size
    finally:
        if image:
            image.close()
```

When you **open** the file an Image object is returned, since `open` is defined as a function in the same file. The image object however contains no method `close`. I guess you wanted to use python's internal `open` function. I don't know what might solve the problem:

* Rename your function `open` to something else
* Remove the `finally` block

Thx in advance.

Cheers,

Peter

----------------------------------------
- Bitbucket: https://bitbucket.org/blockdiag/blockdiag/issue/63
- Originally reported by: Peter Petersky
- Originally created at: 2014-11-18T09:54:33.714

**Comment by tk0miya** [comment]
_From Peter Petersky on 2014-11-18 09:10:41+00:00_

I've got this working by simple changing the finally block from:

```
if image:
    image.close()
```

to 

```
if image:
   del image
```

Now the whole functions looks like this:

```
def get_image_size(filename):
    image = None
    try:
        image = open(filename)
        return image.size
    finally:
        if image:
            del image
```

What do you think?

**Comment by tk0miya** [comment]
_From Takeshi KOMIYA on 2014-11-20 08:31:39+00:00_

Thank you for notification. I try to reproduce that later (maybe, in this weekend).

Just let you know, `get_image_size()` has changed after v1.4.7 release.
It does not call `close()` if the argument is instance of `Image`.
Perhaps it might help you.

BTW, I think it is ugly that these imaging interface that does not have `close()` interface.
I will rewrite `images.open()` and related codes in near future.

**Comment by tk0miya** [comment]
_From Peter Petersky on 2014-11-20 08:51:17+00:00_

Actually I'm using the v1.4.7:

```
# pip freeze | grep blockdiag
blockdiag==1.4.7
blockdiagcontrib-cisco==0.1.8
blockdiagcontrib-class==0.1.3
blockdiagcontrib-math==0.9.0
blockdiagcontrib-octicons==0.1.0
blockdiagcontrib-tex==0.2.0
```

Is there any newer version?

**Comment by tk0miya** [comment]
_From Takeshi KOMIYA on 2014-11-20 11:54:58+00:00_

The fix is not released yet.
I will released v1.5.0 including it soon.

**Comment by tk0miya** [comment]
_From Takeshi KOMIYA on 2014-11-27 03:03:19+00:00_

Fix #63 AttributeError on get_image_size(); Pillow (<= 2.4.x) does not have Image#close()

→ \<\<cset aac49b8b023b660d54a6abb22594135fa92d5dd0\>\>

**Comment by tk0miya** [comment]
_From Takeshi KOMIYA on 2014-11-27 03:07:48+00:00_

Thanks. It was bug with Pillow 2.4.x or older (`Image#close()` was introduced after Pillow 2.5.0).
